lib/server.js

```diff
@@ -283,9 +287,10 @@ class Server extends KarmaEventEmitter {
       }
     })
 
-    this.on('stop', function (done) {
+    this.on('stop', async function (done) {
       this.log.debug('Received stop event, exiting.')
-      return disconnectBrowsers().then(done)
+      await disconnectBrowsers()
+      done()
```

Comment:
The nodejs EventEmitter `on()` method does not AFAICT deal with promises. Its contract is "call `done()`".  The previous code was misleading: returning a promise has no effect.

We could either remove the `return` keyword or we could overwrite EventEmitter.on in the KarmaEventEmitter to check if the function argument is async. If yes, then we send EventEmitter a handler that waits on the return of the async arg and calls done(); the code here then is just `this.on('stop', disconnectBrowsers)`.  If no, then forward the function to EventEmitter as before.

Reply:
I believe `done` callback is configured here:
https://github.com/karma-runner/karma/blob/b99f03fcb8242dd2fd1ac769642c41314679833b/lib/events.js#L52-L56
?

I'll try to look a bit deeper into [`emitAsync`](https://github.com/karma-runner/karma/blob/cc2eff27deb680f789afb34577fd337d2ad5dcac/lib/events.js#L46), `disconnectBrowsers` and try to figure out more elegant solution. Maybe [`beforeExit` event](https://nodejs.org/api/process.html#process_event_beforeexit), as you mentioned in another comment would be helpful here.